### PR TITLE
[DELETE] /api/problems/{problemId} archive

### PR DESCRIPTION
**Pull request by snowmancc**

### CODE REVIEW

**Review by Wally5077 — APPROVED** [review]
LGTM

File: Entities/src/main/java/tw/waterball/judgegirl/entities/problem/Problem.java [hunk 1]
```diff
@@ -66,6 +66,9 @@
 
     private String testcaseIOsFileId;
 
+    @NotNull
```
Comment (Johnny850807) [inline Entities/src/main/java/tw/waterball/judgegirl/entities/problem/Problem.java:69]: remove this

also remove the `@NotNull` above the visible property

File: Spring-Boot/Spring-Boot-Problem/src/main/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemController.java [hunk 2]
```diff
@@ -114,6 +115,11 @@ public void patchProblem(@PathVariable int problemId,
                              @RequestBody PatchProblemUseCase.Request request) {
         patchProblemUseCase.execute(request);
     }
+
+    @DeleteMapping(value = "/{problemId}")
+    public void deleteProblem(@PathVariable int problemId) {
```
Comment (Johnny850807) [inline Spring-Boot/Spring-Boot-Problem/src/main/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemController.java:120]: rename to `archiveOrDeleteProblem`

File: Spring-Boot/Spring-Boot-Problem/src/main/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemController.java [hunk 3]
```diff
@@ -114,6 +115,11 @@ public void patchProblem(@PathVariable int problemId,
                              @RequestBody PatchProblemUseCase.Request request) {
         patchProblemUseCase.execute(request);
     }
+
+    @DeleteMapping(value = "/{problemId}")
```
Comment (Johnny850807) [inline Spring-Boot/Spring-Boot-Problem/src/main/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemController.java:119]: @DeleteMapping("/{problemId}")

File: Spring-Boot/Spring-Boot-Problem/src/main/java/tw/waterball/judgegirl/springboot/problem/repositories/MongoProblemRepository.java [hunk 4]
```diff
@@ -154,6 +154,14 @@ public boolean problemExists(int problemId) {
         return mongoTemplate.find(query(where("_id").in(ids)), Problem.class);
     }
 
+    @Override
+    public void deleteProblemById(int problemId) {
```
Comment (Johnny850807) [inline Spring-Boot/Spring-Boot-Problem/src/main/java/tw/waterball/judgegirl/springboot/problem/repositories/MongoProblemRepository.java:158]: rename to `archiveProblemById `

File: Spring-Boot/Spring-Boot-Problem/src/main/java/tw/waterball/judgegirl/springboot/problem/repositories/MongoProblemRepository.java [hunk 5]
```diff
@@ -154,6 +154,14 @@ public boolean problemExists(int problemId) {
...
+        Update update = new Update();
+        Query query = new Query(where("_id").is(problemId));
+        update.set("deleted", true);
```
Comment (Johnny850807) [inline Spring-Boot/Spring-Boot-Problem/src/main/java/tw/waterball/judgegirl/springboot/problem/repositories/MongoProblemRepository.java:161]: rename `deleted` to `archived`

File: Entities/src/main/java/tw/waterball/judgegirl/entities/problem/Problem.java [hunk 6]
```diff
@@ -66,6 +66,9 @@
...
+    private boolean deleted;
```
Comment (Johnny850807) [inline Entities/src/main/java/tw/waterball/judgegirl/entities/problem/Problem.java:70]: rename to `archived`

File: Entities/src/main/java/tw/waterball/judgegirl/entities/problem/Problem.java [hunk 7]
```diff
@@ -138,6 +141,14 @@ public void setOutputMatchPolicyPluginTag(JudgePluginTag outputMatchPolicyPlugin
         this.outputMatchPolicyPluginTag = outputMatchPolicyPluginTag;
     }
 
+    public boolean isDeleted() {
+        return deleted;
+    }
+
+    public void setDeleted(boolean deleted) {
+        this.deleted = deleted;
+    }
```
Comment (Johnny850807) [inline Entities/src/main/java/tw/waterball/judgegirl/entities/problem/Problem.java:150]: isArchived()

`setArchived(boolean archived)`

File: Problem-Service/src/main/java/tw/waterball/judgegirl/problemservice/domain/repositories/ProblemRepository.java [hunk 8]
```diff
@@ -50,4 +50,6 @@
     boolean problemExists(int problemId);
 
     List<Problem> findProblemsByIds(int... problemIds);
+
+    void deleteProblemById(int problemId);
```
Comment (Johnny850807) [inline Problem-Service/src/main/java/tw/waterball/judgegirl/problemservice/domain/repositories/ProblemRepository.java:54]: `archiveProblemById`

File: Problem-Service/src/main/java/tw/waterball/judgegirl/problemservice/domain/usecases/DeleteProblemUseCase.java [hunk 9]
```diff
@@ -0,0 +1,18 @@
+package tw.waterball.judgegirl.problemservice.domain.usecases;
+
+import tw.waterball.judgegirl.problemservice.domain.repositories.ProblemRepository;
+
+import javax.inject.Named;
+
+@Named
+public class DeleteProblemUseCase extends BaseProblemUseCase {
```
Comment (Johnny850807) [inline Problem-Service/src/main/java/tw/waterball/judgegirl/problemservice/domain/usecases/DeleteProblemUseCase.java:8]: You don't need to rename this class

it is `Delete` for sure

File: Spring-Boot/Spring-Boot-Problem/src/test/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemControllerTest.java [hunk 10]
```diff
@@ -367,11 +367,28 @@ void Given_1_ProblemsSaved_WhenGetProblemsByIds_1_2_ShouldRespondThat_1() throws
         problemsShouldHaveIds(actualProblems, 1);
     }
 
+    @Test
+    void GivenOneProblemCreated_WhenDeleteById_ShouldDeleteSuccessfully() throws Exception {
```
Comment (Johnny850807) [inline Spring-Boot/Spring-Boot-Problem/src/test/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemControllerTest.java:371]: `GivenOneProblemCreated_WhenDeleteIt_ShouldArchiveItSuccessfully`

**Review by Johnny850807 — CHANGES_REQUESTED** [review]
REview done.

File: API/Problem-API/src/main/java/tw/waterball/judgegirl/problemapi/views/ProblemView.java [hunk 11]
```diff
@@ -45,6 +45,7 @@
     public String testcaseIOsFileId;
     public List<Testcase> testcases;
     public boolean visible;
+    public boolean deleted;
```
Comment (Johnny850807) [inline API/Problem-API/src/main/java/tw/waterball/judgegirl/problemapi/views/ProblemView.java:48]: rename this to `archived`

File: Spring-Boot/Spring-Boot-Problem/src/test/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemControllerTest.java (hunk 10 above)
Comment (Johnny850807) [reply Spring-Boot/Spring-Boot-Problem/src/test/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemControllerTest.java]: sorry, I mistype this one, should be `GivenOneProblemCreated_WhenArchiveIt_ShouldArchiveItSuccessfully`

File: API/Problem-API/src/main/java/tw/waterball/judgegirl/problemapi/views/ProblemView.java [hunk 12]
```diff
@@ -57,7 +58,8 @@ public static ProblemView toViewModel(Problem problem) {
                 problem.getTags(),
                 problem.getTestcaseIOsFileId(),
                 problem.getTestcases(),
-                problem.getVisible()
+                problem.getVisible(),
+                problem.isDeleted()
```
Comment (Johnny850807) [inline API/Problem-API/src/main/java/tw/waterball/judgegirl/problemapi/views/ProblemView.java:62]: is this correct?

this should be `problem.isArchived()`

File: API/Problem-API/src/main/java/tw/waterball/judgegirl/problemapi/views/ProblemView.java [hunk 13]
```diff
@@ -70,7 +72,8 @@ public static Problem toEntity(ProblemView view) {
                 .tags(requireNonNullElse(view.tags, emptyList()))
                 .testcases(view.testcases)
                 .testcaseIOsFileId(view.testcaseIOsFileId)
-                .filterPluginTags(requireNonNullElse(view.judgeFilterPluginTags, emptyList()));
+                .filterPluginTags(requireNonNullElse(view.judgeFilterPluginTags, emptyList()))
+                .deleted(view.deleted);
```
Comment (Johnny850807) [inline API/Problem-API/src/main/java/tw/waterball/judgegirl/problemapi/views/ProblemView.java:76]: ??? should be archived

**Review by Johnny850807 — CHANGES_REQUESTED** [review]
Review done

**Review by Johnny850807 — CHANGES_REQUESTED** [review]
Review done

**Review by Johnny850807 — APPROVED** [review]
LGTM

**Review by timt1028 — COMMENTED** [review]
LGTM

**Comment by timt1028** [comment]
LGTM

**Pull request merged**